scipy/stats/_rvs_sampling.py

```diff
@@ -168,3 +171,373 @@ def rvs_ratio_uniforms(pdf, umax, vmin, vmax, size=1, c=0, random_state=None):
         i += 1
 
     return np.reshape(x, size1d)
+
+
+class FastNumericalInverse():
+    r"""
+    Object representing a Fast Numerical Inverse of a probability distribution.
+
+    The initializer of `FastNumericalInverse` accepts `dist`, a frozen instance
+    of `scipy.stats.rv_continuous`, and provides an object with methods
+    that approximate `dist.ppf` and `dist.rvs`. For most distributions,
+    these methods are faster than those of `dist` itself.
+
+    Parameters
+    ----------
+    dist : scipy.stats.rv_frozen
+        Frozen distribution for which a fast numerical inverse is desired.
+    tol : float, optional
+        u-error tolerance (see Notes). The default is 1e-12.
+    max_intervals : int, optional
+        Maximum number of intervals in the cubic Hermite spline used to
+        approximate the percent point function. The default is 100000.
+
+    Attributes
+    ----------
+    intervals : int
+        The number of intervals of the interpolant.
+    midpoint_error : float
+        The maximum u-error at an interpolant interval midpoint.
+
+    Notes
+    -----
+    `FastNumericalInverse` approximates the inverse of a continuous
+    statistical distribution's CDF with a cubic Hermite spline.
+
+    As described in [1]_, it begins by evaluating the distribution's PDF and
+    CDF at a mesh of quantiles ``x`` within the distribution's support.
+    It uses the results to fit a cubic Hermite spline ``H`` such that
+    ``H(p) == x``, where ``p`` is the array of percentiles corresponding
+    with the quantiles ``x``. Therefore, the spline approximates the inverse
+    of the distribution's CDF to machine precision at the percentiles ``p``,
+    but typically, the spline will not be as accurate at the midpoints between
+    the percentile points::
+
+        p_mid = (p[:-1] + p[1:])/2
+
+    so the mesh of quantiles is refined as needed to reduce the maximum
+    "u-error"::
+
+        u_error = np.max(np.abs(dist.cdf(H(p_mid)) - p_mid))
+
+    below the specified tolerance `tol`. Refinement stops when the required
+    tolerance is achieved or when the number of mesh intervals after the next
+    refinement could exceed the maximum allowed number `max_intervals`.
+
+    The approximation will not be accurate for the extreme tails of the
+    distribution (e.g. ``H(1e-17)``, ``H(1)``).
+
+    References
+    ----------
+    .. [1] Hörmann, Wolfgang, and Josef Leydold. "Continuous random variate
+           generation by fast numerical inversion." ACM Transactions on
+           Modeling and Computer Simulation (TOMACS) 13.4 (2003): 347-362.
+
+    Examples
+    --------
+    For some distributions, ``dist.ppf`` are ``dist.rvs`` are quite slow.
```

Comment:
I think this is what you meant
```suggestion
    For some distributions, ``dist.ppf`` and ``dist.rvs`` are quite slow.
```

Reply:
Yes, thanks. I'll add this with the next suggestion.